packages/ui-components/src/components/DateRangePicker/useDateRangePicker.js

```diff
@@ -0,0 +1,171 @@
+/*
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ *
+ */
+import { useEffect } from 'react';
+import moment from 'moment';
+import {
+  constrainDate,
+  DEFAULT_MIN_DATE,
+  GRANULARITIES,
+  GRANULARITIES_WITH_ONE_DATE,
+  GRANULARITY_CONFIG,
+  momentToDateString,
+  roundStartEndDates,
+  roundStartDate,
+  roundEndDate,
+  toStandardDateString,
+} from '../Chart';
+
+const DEFAULT_GRANULARITY = GRANULARITY_CONFIG[GRANULARITIES.DAY];
+
+/**
+ *
+ * @param isSingleDate
+ * @param granularity
+ * @param startDate
+ * @param endDate
+ * @returns {*|string}
+ */
+const getDatesAsString = (isSingleDate, granularity, startDate, endDate) => {
+  const { rangeFormat } = GRANULARITY_CONFIG[granularity] || DEFAULT_GRANULARITY;
+  const formattedStartDate = momentToDateString(startDate, granularity, rangeFormat);
+  const formattedEndDate = momentToDateString(endDate, granularity, rangeFormat);
+
+  return isSingleDate ? formattedEndDate : `${formattedStartDate} - ${formattedEndDate}`;
+};
+
+/**
+ *
+ * @param isSingleDate
+ * @param granularity
+ * @param minMomentDate
+ * @param maxMomentDate
+ * @returns {{defaultEndDate: moment.Moment, defaultStartDate: moment.Moment}}
+ */
+const getDefaultDates = (isSingleDate, granularity, minMomentDate, maxMomentDate) => {
+  const defaultStartDate = isSingleDate
+    ? constrainDate(moment(), minMomentDate, maxMomentDate)
+    : minMomentDate;
+
+  const defaultEndDate = isSingleDate ? defaultStartDate : maxMomentDate;
+
+  return {
+    defaultStartDate: roundStartDate(granularity, defaultStartDate),
+    defaultEndDate: roundEndDate(granularity, defaultEndDate),
+  };
+};
+
+/**
+ *
+ * @param granularity
+ * @param startDate
+ * @param endDate
+ * @param defaultStartDate
+ * @param defaultEndDate
+ * @returns {{currentStartDate: (moment.Moment|*), currentEndDate: (moment.Moment|*)}}
+ */
+const getCurrentDates = (granularity, startDate, endDate, defaultStartDate, defaultEndDate) => ({
+  currentStartDate: startDate ? roundStartDate(granularity, startDate) : defaultStartDate,
+  currentEndDate: endDate ? roundEndDate(granularity, endDate) : defaultEndDate,
+});
+
+/**
+ *
+ * @param startDate
+ * @param endDate
+ * @param minDate
+ * @param maxDate
+ * @param granularity
+ * @param onSetDates
+ * @returns {{handleReset: handleReset, handleDateChange: handleDateChange, nextDisabled: boolean, labelText: (*|string), isSingleDate: boolean, currentStartDate: (*|moment.Moment), currentEndDate: (*|moment.Moment), prevDisabled: boolean, changePeriod: changePeriod}}
+ */
+export const useDateRangePicker = ({
+  startDate,
+  endDate,
+  minDate,
+  maxDate,
+  granularity,
+  onSetDates,
+}) => {
+  /**
+   * Call the on change handler prop using iso formatted date
+   */
+  const handleDateChange = (newStartDate, newEndDate) => {
+    onSetDates(toStandardDateString(newStartDate), toStandardDateString(newEndDate));
```

Comment:
Good question. The min/max date validation is handled in the onSubmit handler of the datePicker dialog (I believe there is also server side validation). This is the existing validation and I think it is the best way of doing it too because it could be frustrating for users if they get an error while they are half way through setting their date range options.